File: solutions/lua/diamond/1/diamond.lua
return function(which)
  local order <const> = string.byte(which) - string.byte('A')
  if order < 0 or 25 < order then
    error("input is out of valid range")
  end
  local side <const> = 2 * order + 1
  local buf <const> = {}
  for i = 1, side do
    buf[i] = ' '
  end
  local diamond <const> = {}
  local top = 1
  local bottom = side
  local left = order + 1
  local right = left
  for o = 0, order do
    local char <const> = string.char(o + string.byte('A'))
    buf[left], buf[right] = char, char
    diamond[top] = table.concat(buf)
    diamond[bottom] = diamond[top]
    buf[left], buf[right] = ' ', ' '
    left, right = left - 1, right + 1
    top, bottom = top + 1, bottom - 1
  end
  return table.concat(diamond, '\n') .. '\n'
end

File: solutions/csharp/diamond/1/Diamond.cs
public static class Diamond
{
    public static string Make(char target)
    {
        int order = target - 'A';
        if (order < 0 || order > 25)
            throw new Exception("target must be between A and Z");
        int side = 2 * order + 1;
        int left = order, right = order, top = 0, bottom = side - 1;
        char[] buf = new char[side];
        Array.Fill(buf, ' ');
        string[] diamond = new string[side];
        for (char ch = 'A'; ch <= target; ++ch)
        {
            buf[left] = buf[right] = ch;
            diamond[top++] = diamond[bottom--] = new String(buf);
            buf[left--] = buf[right++] = ' ';
        }
        return String.Join('\n', diamond);
    }
}